lib/Runtime/Provisioner/Provisioner.cpp

```diff
@@ -75,14 +76,19 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module) {
     uint64_t totalMemory = 0;
     FunctionMapTy functionMap;
     for (auto &node : device.second) {
-      Function *function = module.getFunction(node->name);
-      CompilationOptions compileOptions;
-      compileOptions.collectConstants = false;
-      auto compiled = backend_->compile(function, compileOptions);
-      node->runtimeBundle =
-          llvm::make_unique<RuntimeBundle>(compiled->getRuntimeBundle());
-      functionMap.emplace(node->name, compiled.get());
-      functions_.emplace(node->name, std::move(compiled));
+      // Only compile if we haven't compiled before. If we have previously
+      // compiled the function reuse it.
+      auto it = functions_.find(node->name);
+      if (it == functions_.end()) {
+        Function *function = module.getFunction(node->name);
+        CompilationOptions compileOptions;
+        compileOptions.collectConstants = false;
```

Comment:
This is not clear to me by reading the code what the provisioner is doing and why it needs to know something about collecting constants.

Reply:
I agree. This is an existing fact of the provisioner rather than something introduced here though. I think the underlying issue is that runtime bundle isn't actually a thing that's needed by all backends, only backends running in the same memory space as the compiler (CPU, GPU, Interpreter) so while collect constants part of it has been moved to those devicemanagers, the runtime bundle still remains in the runtime but could probably be moved there as well. This is a conversation bigger than this pr though.